File: app/src/main/java/com/viafourasample/src/activities/login/LoginActivity.java
package com.viafourasample.src.activities.login;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.loginradius.androidsdk.helper.LoginRadiusSDK;
import com.loginradius.androidsdk.resource.SocialProviderConstant;
import com.viafoura.sampleapp.R;
import com.viafourasample.src.activities.signup.SignUpActivity;
import com.viafourasample.src.managers.ColorManager;
import com.viafourasample.src.model.SettingKeys;
import com.viafourasample.src.utils.AndroidUtils;
import com.viafourasdk.src.model.local.VFDefaultColors;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

public class LoginActivity extends AppCompatActivity {
    private LoginViewModel viewModel = new LoginViewModel();

    private static int SIGNUP_RESULT_CODE = 1;
    private static int LOGINRADIUS_RESULT_CODE = 2;

    private TextInputEditText emailText, passwordText;
    private TextInputLayout emailLayout, passwordLayout;
    private TextView resetText, signupText;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        getSupportActionBar().setTitle("Log-in");

        if(ColorManager.isDarkMode(getApplicationContext())){
            findViewById(R.id.login_holder).setBackgroundColor(ContextCompat.getColor(getApplicationContext(), R.color.colorBackgroundArticle));
        }

        ((ProgressBar) findViewById(R.id.login_loading)).getIndeterminateDrawable().setColorFilter(
                ContextCompat.getColor(getApplicationContext(), R.color.colorPrimary),
                android.graphics.PorterDuff.Mode.SRC_IN);

        setupTextInputs();

        findViewById(R.id.login_facebook).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                LoginRadiusSDK.NativeLogin nativeLogin = new LoginRadiusSDK.NativeLogin();
                nativeLogin.startFacebookNativeLogin(LoginActivity.this, LOGINRADIUS_RESULT_CODE);
            }
        });

        findViewById(R.id.login_linkedin).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                launchSocialLogin(SocialProviderConstant.LINKEDIN);
            }
        });

        findViewById(R.id.login_twitter).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                launchSocialLogin(SocialProviderConstant.TWITTER);
            }
        });

        findViewById(R.id.login_google).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                LoginRadiusSDK.NativeLogin nativeLogin = new LoginRadiusSDK.NativeLogin();
                nativeLogin.startGoogleNativeLogin(LoginActivity.this, LOGINRADIUS_RESULT_CODE);
            }
        });

        setupReset();
        setupSignup();

        findViewById(R.id.login_submit).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String email = emailText.getText().toString();
                String password = passwordText.getText().toString();

                findViewById(R.id.login_submit).setVisibility(View.GONE);
                findViewById(R.id.login_loading).setVisibility(View.VISIBLE);

                viewModel.login(email, password, new LoginViewModel.LoginCallback() {
                    @Override
                    public void onSuccess() {
                        findViewById(R.id.login_submit).setVisibility(View.VISIBLE);
                        findViewById(R.id.login_loading).setVisibility(View.GONE);

                        onBackPressed();
                    }

                    @Override
                    public void onError(String errorMessage) {
                        findViewById(R.id.login_submit).setVisibility(View.VISIBLE);
                        findViewById(R.id.login_loading).setVisibility(View.GONE);

                        showAlert("Invalid credentials", errorMessage);
                    }
                });
            }
        });
    }

    private void showAlert(String title, String message){
        new AlertDialog.Builder(LoginActivity.this)
                .setTitle(title)
                .setIcon(null)
                .setMessage(message)
                .setPositiveButton(android.R.string.yes, null)
                .setNegativeButton(android.R.string.no, null)
                .setIcon(android.R.drawable.ic_dialog_alert)
                .show();
    }

    private void launchSocialLogin(SocialProviderConstant socialProvider){
        LoginRadiusSDK.WebLogin webLogin = new LoginRadiusSDK.WebLogin();
        webLogin.setProvider(socialProvider);
        webLogin.startWebLogin(LoginActivity.this, LOGINRADIUS_RESULT_CODE);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if(requestCode == SIGNUP_RESULT_CODE){
            if(resultCode == Activity.RESULT_OK){
                onBackPressed();
            }
        } else if(requestCode == LOGINRADIUS_RESULT_CODE){
            if (data != null){
                String accessToken = data.getStringExtra("accesstoken");
                String provider = data.getStringExtra("provider");

                viewModel.socialLogin(accessToken, new LoginViewModel.SocialLoginCallback() {
                    @Override
                    public void onSuccess() {

                    }

                    @Override
                    public void onError() {

                    }
                });
            }
        }
    }

    private void setupSignup(){
        signupText = findViewById(R.id.login_signup);
        signupText.setTextColor(ColorManager.isDarkMode(this) ? Color.WHITE : Color.BLACK);
        signupText.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(getApplicationContext(), SignUpActivity.class);
                startActivityForResult(intent, SIGNUP_RESULT_CODE);
            }
        });
    }

    private void setupReset(){
        resetText = findViewById(R.id.login_reset);
        resetText.setTextColor(ColorManager.isDarkMode(this) ? Color.WHITE : Color.BLACK);
        resetText.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                final EditText emailEditText = new EditText(LoginActivity.this);
                emailEditText.setSingleLine();

                FrameLayout container = new FrameLayout(LoginActivity.this);
                FrameLayout.LayoutParams params = new  FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
                params.leftMargin = (int) AndroidUtils.convertDpToPixel(20, LoginActivity.this);
                params.rightMargin = (int) AndroidUtils.convertDpToPixel(20, LoginActivity.this);
                emailEditText.setLayoutParams(params);
                container.addView(emailEditText);

                AlertDialog dialog = new AlertDialog.Builder(LoginActivity.this)
                        .setTitle("Reset password")
                        .setMessage("Enter your e-mail")
                        .setView(container)
                        .setPositiveButton("Submit", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                String email = String.valueOf(emailEditText.getText());
                                viewModel.resetPassword(email, new LoginViewModel.PasswordResetCallback() {
                                    @Override
                                    public void onSuccess() {

                                    }

                                    @Override
                                    public void onError() {

                                    }
                                });
                                dialog.dismiss();
                            }
                        })
                        .setNegativeButton("Cancel", null)
                        .create();
                dialog.show();
            }
        });
    }

    private void setupTextInputs(){
        emailText = findViewById(R.id.login_email_text);
        passwordText = findViewById(R.id.login_password_text);

        emailText.setTextColor(ColorManager.isDarkMode(this) ? Color.WHITE : Color.BLACK);
        passwordText.setTextColor(ColorManager.isDarkMode(this) ? Color.WHITE : Color.BLACK);

        passwordLayout = findViewById(R.id.login_password);
        emailLayout = findViewById(R.id.login_email);

        setInputTextLayoutColor(ColorManager.isDarkMode(this) ? Color.WHITE : Color.BLACK, emailLayout);
        setInputTextLayoutColor(ColorManager.isDarkMode(this) ? Color.WHITE : Color.BLACK, passwordLayout);
    }

    private void setInputTextLayoutColor(int color, TextInputLayout textInputLayout) {
        try {
            Field field = textInputLayout.getClass().getDeclaredField("focusedTextColor");
            field.setAccessible(true);
            int[][] states = new int[][]{
                    new int[]{}
            };
            int[] colors = new int[]{
                    color
            };
            ColorStateList myList = new ColorStateList(states, colors);
            field.set(textInputLayout, myList);

            Field fDefaultTextColor = TextInputLayout.class.getDeclaredField("defaultHintTextColor");
            fDefaultTextColor.setAccessible(true);
            fDefaultTextColor.set(textInputLayout, myList);

            Method method = textInputLayout.getClass().getDeclaredMethod("updateLabelState", boolean.class);
            method.setAccessible(true);
            method.invoke(textInputLayout, true);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/viafourasample/src/activities/login/LoginActivity.kt
package com.viafourasample.src.activities.login

import android.app.Activity
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.PorterDuff
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.viafoura.sampleapp.R
import com.viafourasample.src.activities.signup.SignUpActivity
import com.viafourasample.src.managers.ColorManager
import com.viafourasample.src.utils.AndroidUtils
import com.viafourasample.src.utils.InsetsUtils
import com.viafourasdk.src.utils.VFInsetsUtils

class LoginActivity : AppCompatActivity() {
    private val viewModel = LoginViewModel()

    private lateinit var emailText: TextInputEditText
    private lateinit var passwordText: TextInputEditText
    private lateinit var emailLayout: TextInputLayout
    private lateinit var passwordLayout: TextInputLayout
    private lateinit var resetText: TextView
    private lateinit var signupText: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        InsetsUtils.applyActionBarInsets(this)
        VFInsetsUtils.applyBottomInsetsWithIme(findViewById<View>(R.id.login_holder))

        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
        supportActionBar!!.title = "Log-in"

        if (ColorManager.isDarkMode(applicationContext)) {
            findViewById<View>(R.id.login_holder).setBackgroundColor(
                ContextCompat.getColor(applicationContext, R.color.colorBackgroundArticle)
            )
        }

        findViewById<ProgressBar>(R.id.login_loading).indeterminateDrawable.setColorFilter(
            ContextCompat.getColor(applicationContext, R.color.colorPrimary),
            PorterDuff.Mode.SRC_IN
        )

        setupTextInputs()

        setupReset()
        setupSignup()

        findViewById<View>(R.id.login_submit).setOnClickListener {
            val email = emailText.text.toString()
            val password = passwordText.text.toString()

            findViewById<View>(R.id.login_submit).visibility = View.GONE
            findViewById<View>(R.id.login_loading).visibility = View.VISIBLE

            viewModel.login(email, password, object : LoginViewModel.LoginCallback {
                override fun onSuccess() {
                    findViewById<View>(R.id.login_submit).visibility = View.VISIBLE
                    findViewById<View>(R.id.login_loading).visibility = View.GONE

                    onBackPressed()
                }

                override fun onError(errorMessage: String?) {
                    findViewById<View>(R.id.login_submit).visibility = View.VISIBLE
                    findViewById<View>(R.id.login_loading).visibility = View.GONE

                    showAlert("Invalid credentials", errorMessage)
                }
            })
        }
    }

    private fun showAlert(title: String, message: String?) {
        AlertDialog.Builder(this@LoginActivity)
            .setTitle(title)
            .setIcon(null)
            .setMessage(message)
            .setPositiveButton(android.R.string.yes, null)
            .setNegativeButton(android.R.string.no, null)
            .setIcon(android.R.drawable.ic_dialog_alert)
            .show()
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == SIGNUP_RESULT_CODE) {
            if (resultCode == Activity.RESULT_OK) {
                onBackPressed()
            }
        }
    }

    private fun setupSignup() {
        signupText = findViewById(R.id.login_signup)
        signupText.setTextColor(if (ColorManager.isDarkMode(this)) Color.WHITE else Color.BLACK)
        signupText.setOnClickListener {
            val intent = Intent(applicationContext, SignUpActivity::class.java)
            startActivityForResult(intent, SIGNUP_RESULT_CODE)
        }
    }

    private fun setupReset() {
        resetText = findViewById(R.id.login_reset)
        resetText.setTextColor(if (ColorManager.isDarkMode(this)) Color.WHITE else Color.BLACK)
        resetText.setOnClickListener {
            val emailEditText = EditText(this@LoginActivity)
            emailEditText.setSingleLine()

            val container = FrameLayout(this@LoginActivity)
            val params = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            params.leftMargin = AndroidUtils.convertDpToPixel(20f, this@LoginActivity).toInt()
            params.rightMargin = AndroidUtils.convertDpToPixel(20f, this@LoginActivity).toInt()
            emailEditText.layoutParams = params
            container.addView(emailEditText)

            val dialog = AlertDialog.Builder(this@LoginActivity)
                .setTitle("Reset password")
                .setMessage("Enter your e-mail")
                .setView(container)
                .setPositiveButton("Submit") { dialog, _ ->
                    val email = emailEditText.text.toString()
                    viewModel.resetPassword(email, object : LoginViewModel.PasswordResetCallback {
                        override fun onSuccess() {
                        }

                        override fun onError() {
                        }
                    })
                    dialog.dismiss()
                }
                .setNegativeButton("Cancel", null)
                .create()
            dialog.show()
        }
    }

    private fun setupTextInputs() {
        emailText = findViewById(R.id.login_email_text)
        passwordText = findViewById(R.id.login_password_text)

        emailText.setTextColor(if (ColorManager.isDarkMode(this)) Color.WHITE else Color.BLACK)
        passwordText.setTextColor(if (ColorManager.isDarkMode(this)) Color.WHITE else Color.BLACK)

        passwordLayout = findViewById(R.id.login_password)
        emailLayout = findViewById(R.id.login_email)

        setInputTextLayoutColor(
            if (ColorManager.isDarkMode(this)) Color.WHITE else Color.BLACK,
            emailLayout
        )
        setInputTextLayoutColor(
            if (ColorManager.isDarkMode(this)) Color.WHITE else Color.BLACK,
            passwordLayout
        )
    }

    private fun setInputTextLayoutColor(color: Int, textInputLayout: TextInputLayout) {
        try {
            val field = textInputLayout.javaClass.getDeclaredField("focusedTextColor")
            field.isAccessible = true
            val states = arrayOf(intArrayOf())
            val colors = intArrayOf(color)
            val myList = ColorStateList(states, colors)
            field.set(textInputLayout, myList)

            val fDefaultTextColor =
                TextInputLayout::class.java.getDeclaredField("defaultHintTextColor")
            fDefaultTextColor.isAccessible = true
            fDefaultTextColor.set(textInputLayout, myList)

            val method = textInputLayout.javaClass.getDeclaredMethod(
                "updateLabelState",
                Boolean::class.javaPrimitiveType
            )
            method.isAccessible = true
            method.invoke(textInputLayout, true)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
        }

        return super.onOptionsItemSelected(item)
    }

    companion object {
        private const val SIGNUP_RESULT_CODE = 1
    }
}
